File: dataModule/src/main/java/com/mobnetic/coinguardian/model/market/BtcMarkets.java
package com.mobnetic.coinguardian.model.market;

import java.util.HashMap;
import java.util.LinkedHashMap;

import org.json.JSONObject;

import com.mobnetic.coinguardian.model.CheckerInfo;
import com.mobnetic.coinguardian.model.Market;
import com.mobnetic.coinguardian.model.Ticker;
import com.mobnetic.coinguardian.model.currency.Currency;
import com.mobnetic.coinguardian.model.currency.VirtualCurrency;

public class BtcMarkets extends Market {

	private final static String NAME = "BtcMarkets.net";
	private final static String TTS_NAME = "BTC Markets net";
	private final static String URL = "https://api.btcmarkets.net/market/%1$s/%2$s/tick";
	private final static HashMap<String, CharSequence[]> CURRENCY_PAIRS = new LinkedHashMap<String, CharSequence[]>();
	static {
		CURRENCY_PAIRS.put(VirtualCurrency.BTC, new String[]{
				Currency.AUD
			});
		CURRENCY_PAIRS.put(VirtualCurrency.LTC, new String[]{
				VirtualCurrency.BTC,
				Currency.AUD
			});
		CURRENCY_PAIRS.put(VirtualCurrency.ETC, new String[]{
				VirtualCurrency.BTC,
				Currency.AUD
			});
		CURRENCY_PAIRS.put(VirtualCurrency.ETH, new String[]{
				VirtualCurrency.BTC,
				Currency.AUD
			});
		CURRENCY_PAIRS.put(VirtualCurrency.XRP, new String[]{
				VirtualCurrency.BTC,
				Currency.AUD
			});
		CURRENCY_PAIRS.put(VirtualCurrency.BCH, new String[]{
				VirtualCurrency.BTC,
				Currency.AUD
			});
	}
	
	public BtcMarkets() {
		super(NAME, TTS_NAME, CURRENCY_PAIRS);
	}
	
	@Override
	public String getUrl(int requestId, CheckerInfo checkerInfo) {
		return String.format(URL, checkerInfo.getCurrencyBase(), checkerInfo.getCurrencyCounter());
	}
	
	@Override
	protected void parseTickerFromJsonObject(int requestId, JSONObject jsonObject, Ticker ticker, CheckerInfo checkerInfo) throws Exception {
		ticker.bid = jsonObject.getDouble("bestBid");
		ticker.ask = jsonObject.getDouble("bestAsk");
		ticker.last = jsonObject.getDouble("lastPrice");
		ticker.timestamp = jsonObject.getLong("timestamp");
	}
}

File: dataModule/src/main/java/com/aneonex/bitcoinchecker/datamodule/model/market/BtcMarkets.kt
package com.aneonex.bitcoinchecker.datamodule.model.market

import com.aneonex.bitcoinchecker.datamodule.model.CheckerInfo
import com.aneonex.bitcoinchecker.datamodule.model.CurrencyPairInfo
import com.aneonex.bitcoinchecker.datamodule.model.Ticker
import com.aneonex.bitcoinchecker.datamodule.model.market.generic.SimpleMarket
import com.aneonex.bitcoinchecker.datamodule.util.TimeUtils
import com.aneonex.bitcoinchecker.datamodule.util.forEachJSONObject
import org.json.JSONArray
import org.json.JSONObject

class BtcMarkets : SimpleMarket(
        "BtcMarkets.net",
        "https://api.btcmarkets.net/v3/markets",
        "https://api.btcmarkets.net/v3/markets/%1\$s/ticker",
        "BTC Markets net"
) {
    override fun getPairId(checkerInfo: CheckerInfo): String {
        // Compatibility with old app version
        return checkerInfo.currencyPairId ?: "${checkerInfo.currencyBase}-${checkerInfo.currencyCounter}"
    }

    override fun parseCurrencyPairs(requestId: Int, responseString: String, pairs: MutableList<CurrencyPairInfo>) {
        JSONArray(responseString).forEachJSONObject { market ->
            pairs.add(CurrencyPairInfo(
                    market.getString("baseAssetName"),
                    market.getString("quoteAssetName"),
                    market.getString("marketId")
            ))
        }
    }

    @Throws(Exception::class)
    override fun parseTickerFromJsonObject(requestId: Int, jsonObject: JSONObject, ticker: Ticker, checkerInfo: CheckerInfo) {
        ticker.apply {
            last = jsonObject.getDouble("lastPrice")

            bid = jsonObject.getDouble("bestBid")
            ask = jsonObject.getDouble("bestAsk")

            low = jsonObject.getDouble("low24h")
            high = jsonObject.getDouble("high24h")

            vol = jsonObject.getDouble("volume24h")
            volQuote = jsonObject.getDouble("volumeQte24h")

            timestamp = TimeUtils.convertISODateToTimestamp(jsonObject.getString("timestamp"))
        }
    }
}
